mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,311 @@ def load_model(model_uri):
+          but, the _autolog_training_status is a global status without thread-safe lock protecting.
+          This safe guarding will prevent code run into this case.
+        """
+        return _SklearnTrainingSession.is_active() and self._log_post_training_metrics_enabled
+
+    def disable_log_post_training_metrics(self):
+        old_status = self._log_post_training_metrics_enabled
+
+        class LogPostTrainingMetricsDisabledScope:
+            def __enter__(inner_self):
+                self._log_post_training_metrics_enabled = False
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self._log_post_training_metrics_enabled = old_status
+
+        return LogPostTrainingMetricsDisabledScope()
+
+    def get_run_id_for_model(self, model):
+        if hasattr(model, '_mlflow_uuid'):
+            return self._model_id_to_run_id_map.get(model._mlflow_uuid, None)
+        return None
+
+    @staticmethod
+    def get_model_id(model):
+        return model._mlflow_uuid
+
+    @staticmethod
+    def is_metrics_value_loggable(metric_value):
+        return np.isscalar(metric_value)
+
+    def register_model(self, model, run_id):
+        """
+        In `patched_fit`, we need register the model with the run_id used in `patched_fit`
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        model._mlflow_uuid = uuid4()
+        self._model_id_to_run_id_map[model._mlflow_uuid] = run_id
+
+    @staticmethod
+    def gen_name_with_index(name, index):
+        assert index >= 0
+        if index == 0:
+            return name
+        else:
+            # Use '-' as the separator between name and index,
+            # The '-' is not valid character in python var name
+            # so it can prevent name conflicts after appending index.
+            return f'{name}-{index + 1}'
+
+    @staticmethod
+    def gen_dataset_sample_rows(eval_dataset):
+        n_sample = max(len(eval_dataset), INPUT_EXAMPLE_SAMPLE_ROWS)
+        return eval_dataset[:n_sample]
+
+    def register_eval_dataset(self, model, eval_dataset):
+        """
+        Register eval dataset into eval_dataset_info_map, it will do:
+         1. inspect eval dataset var name.
+         2. check whether eval_dataset_info_map already registered this eval dataset.
+            will check by object id and comparing row samples.
+            Note: only check object id is not sufficient, id may be reused.
+         3. register eval dataset (with id and row samples information)
+         4. return eval dataset name with index.
+        """
+        eval_dataset_name = _inspect_original_var_name(eval_dataset)
+        eval_dataset_id = id(eval_dataset)
+        if eval_dataset_name is None:
+            eval_dataset_name = 'unknown_dataset'
+
+        model_id = self.get_model_id(model)
+        registered_dataset_list = self._eval_dataset_info_map[model_id][eval_dataset_name]
```

Comment:
Still not sure why we need `model_id` as an intermediate layer of indirection. I understand that we map prediction results to tuples of (eval_dataset_name, model_id), but it seems like the only thing we do with that `model_id` afterwards is map it to a `run_id`. Can't we just map the prediction results to `run_id` directly and add the run id as an attribute of the model? The model is always accessible in `predict()`.

Reply:
> Can't we just map the prediction results to run_id directly 

If soseems make sense. I will try.